File: hermetic_tests/go/greeter_test.go
package greeter

import (
	"runtime"
	"testing"
)

func TestGreet(t *testing.T) {
	got := Greet("World")
	want := "Hello, World!"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestHermeticToolchain(t *testing.T) {
	// rules_go pins the Go SDK in MODULE.bazel (go_sdk.download version = "1.24.5").
	// If this test is compiled with any other Go version, it fails — proving the
	// hermetic toolchain is in use regardless of what Go is installed on the host.
	const want = "go1.24.5"
	if got := runtime.Version(); got != want {
		t.Errorf("Go version = %q, want %q — hermetic toolchain not in use", got, want)
	}
}

File: hermetic_tests/cc/greeter_test.cc
#include "hermetic_tests/cc/greeter.h"

#include "gtest/gtest.h"

TEST(GreeterTest, Greets) {
  EXPECT_EQ(Greet("World"), "Hello, World!");
}

TEST(GreeterTest, HermeticToolchain) {
  // toolchains_llvm pins the LLVM version in tools/cpp.MODULE.bazel.
  // Linux x86_64 (RBE) uses LLVM 15.0.6; macOS uses LLVM 18.1.8.
  // Both are >= 15. If compiled with an older system compiler this fails,
  // proving the hermetic LLVM toolchain is in use.
  EXPECT_GE(__clang_major__, 15)
      << "Compiled with Clang " << __clang_major__
      << " — expected >= 15 from toolchains_llvm";
}
